File: orquestator/handlers/gather.go
package handlers

import (
	"log"
	"net/http"
	"orquestator/actions"
	"orquestator/api/session"

	"github.com/gin-gonic/gin"
)

// GatherHandler handles the speech transcription and voice logic
func GatherHandler(context *gin.Context) {
	// Use context.PostForm("From") to get the client number
	//fromNumber := context.PostForm("From")
	speech := context.PostForm("SpeechResult")
	callSid := context.PostForm("CallSid")

	log.Printf("(%s): %s\n", callSid, speech)

	response, err := session.Send(callSid, speech)
	if err != nil {
		log.Println("Agent Send Error:", err)
		// helper.EndCall("Lo siento, tuvimos un error interno, por favor intentalo más tarde.")
	}

	log.Printf("Action: %s", response.Action)

	// Get the action handler
	actionHandler := actions.GetActionHandler(response.Action)
	if actionHandler == nil {
		log.Println("Error action not found:", response.Action)
		return
	}

	// Get the action result
	result, err := actionHandler(callSid, response.Data)
	if err != nil {
		log.Println("Error handling action:", response.Action, err)
		context.String(http.StatusInternalServerError, err.Error())
		return
	}

	// Send result to Twilio
	context.Header("Content-Type", "text/xml")
	context.String(http.StatusOK, result)
}

File: orquestator/handlers/gather.py
from fastapi import APIRouter, Form
from fastapi.responses import Response
import logging

from api.session import send_message
from actions.registry import get_action_handler

router = APIRouter()
logger = logging.getLogger(__name__)

@router.post("/gather")
async def gather_handler(
    From: str = Form(...),
    SpeechResult: str = Form(...),
    CallSid: str = Form(...)
):
    try:
        logger.info(f"({CallSid}): {SpeechResult}")
        
        response = await send_message(CallSid, SpeechResult)
        logger.info(f"Action: {response['action']}")
        
        handler = get_action_handler(response["action"])
        if not handler:
            logger.error(f"Action not found: {response['action']}")
            return Response(content="", media_type="text/xml")
        
        result = await handler(CallSid, From, response["data"])
        
        return Response(content=result, media_type="text/xml")
    
    except Exception as e:
        logger.error(f"Gather error: {e}")
        return Response(content="", media_type="text/xml")